tests/integration/general/test_snapcraftctl_set_version.py

```diff
@@ -0,0 +1,76 @@
+# -*- Mode:Python; indent-tabs-mode:nil; tab-width:4 -*-
+#
+# Copyright (C) 2016-2018 Canonical Ltd
+#
+# This program is free software: you can redistribute it and/or modify
+# it under the terms of the GNU General Public License version 3 as
+# published by the Free Software Foundation.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+import os
+import subprocess
+import textwrap
+import yaml
+
+from testtools.matchers import Equals, Contains
+
+from tests import integration
+
+
+class SnapcraftctlSetVersionTestCase(integration.TestCase):
+
+    def test_set_version(self):
+        self.construct_yaml(
+            version=None, adopt_info='my-part', parts=textwrap.dedent("""\
+                my-part:
+                  plugin: nil
+                  override-pull: snapcraftctl set-version override-version
+                """))
+
+        self.run_snapcraft('prime')
+
+        with open(os.path.join('prime', 'meta', 'snap.yaml')) as f:
+            y = yaml.load(f)
+
+        self.assertThat(y['version'], Equals('override-version'))
+
+    def test_set_version_no_overwrite(self):
+        self.construct_yaml(
+            version='test-version', adopt_info='my-part',
+            parts=textwrap.dedent("""\
+                my-part:
+                  plugin: nil
+                  override-pull: snapcraftctl set-version override-version
+                """))
+
+        self.run_snapcraft('prime')
+
+        with open(os.path.join('prime', 'meta', 'snap.yaml')) as f:
+            y = yaml.load(f)
+
+        self.assertThat(y['version'], Equals('test-version'))
```

Comment:
this should actually be the opposite

Reply:
This is testing the case when a version is specified in the snapcraft.yaml, which takes ultimate precedence unless it's missing. The metadata extractors work that way, so I duplicated that for scriptlets. Otherwise it's possible to be using a part that gives you e.g. an icon, but a version you don't want.

Reply:
this is the other way around to what we already have with `version-script` where you can do

```bash
if [ $SNAPCRAFT_PROJECT_GRADE = 'stable' ]; then
    echo $SNAPCRAFT_PROJECT_VERSION
else
    echo $SNAPCRAFT_PROJECT_VERSION-unstable
fi
```

Reply:
If you feel strongly about this, we should warn about all the properties we are ignoring because they are already set in YAML.